### PR TITLE
Added Arduino Nano ESP32-S3

### PR DESCRIPTION
**Pull request by makermelissa**
Depends on https://github.com/adafruit/nvm.toml/pull/15.

I'm still testing, but most things seem to work so far. 

One issue is there are some board with inverted green/blue on the status RGB LEDs (see https://support.arduino.cc/hc/en-us/articles/9589073738012-About-Nano-ESP32-boards-with-inverted-green-and-blue-pins). Mine happens to be one of the ones with inverted green and blue pins, but I created this board based on the non-inverted LED.

Question is:should we have a second board def for inverted led or just be fine with one board definition?

### CODE REVIEW

**Comment by tannewt** [comment]
> Question is:should we have a second board def for inverted led or just be fine with one board definition?

I think two defs are probably best to cover everyone.

**Comment by makermelissa** [comment]
Thanks

**Comment by makermelissa** [comment]
Ok, I've added a board with the inverted status led. I kept having it crash whenever I was trying to access the TX/RX pins. Upon investigation, I realized that these are actually the RXD and TXD pins for the debug interface, so set them as such. 

I'm not 100% sure that I set up to use the external flash correctly either. It matches other boards in the parameters used and the external flash file is looking for the parameters I used, but I'm not getting the full 16MB of space. It may just a FAT12 thing, but that should be able to get up to 16MB.

**Comment by makermelissa** [comment]
Oh yeah, it's failing duplicate VID/PID now because of the 2 board defs for the same board.

**Comment by tannewt** [comment]
> I'm not 100% sure that I set up to use the external flash correctly either. It matches other boards in the parameters used and the external flash file is looking for the parameters I used, but I'm not getting the full 16MB of space. It may just a FAT12 thing, but that should be able to get up to 16MB.

How much memory are you seeing?

> Oh yeah, it's failing duplicate VID/PID now because of the 2 board defs for the same board.

You can exempt it here: https://github.com/adafruit/circuitpython/blob/main/tools/ci_check_duplicate_usb_vid_pid.py#L33

**Comment by makermelissa** [comment]
> > I'm not 100% sure that I set up to use the external flash correctly either. It matches other boards in the parameters used and the external flash file is looking for the parameters I used, but I'm not getting the full 16MB of space. It may just a FAT12 thing, but that should be able to get up to 16MB.
> 
> How much memory are you seeing?

5.9MB, which now that I think about it more, probably means it's using internal flash rather than external.

**Comment by tannewt** [comment]
> > How much memory are you seeing?
> 
> 5.9MB, which now that I think about it more, probably means it's using internal flash rather than external.

Did you try erasing the filesystem? I think this is how much you'd get if it was configured for 8MB flash. (There isn't internal flash so the remaining 1MB or so is for CircuitPython.)

**Comment by makermelissa** [comment]
No, I didn't try that. I'll give it a go.

**Comment by makermelissa** [comment]
Same size and I just realized it's actually fat16.

**Comment by makermelissa** [comment]
Also, it has 8mb internal flash and an additional 16mb external flash. Arduino just used all the internal flash for the bootloader I think.

**Comment by tannewt** [comment]
> Also, it has 8mb internal flash and an additional 16mb external flash. Arduino just used all the internal flash for the bootloader I think.

Oh weird! The module has 8MB qspi plus 16MB externally. Do you know which flash is used by micropython?

**Comment by makermelissa** [comment]
> Oh weird! The module has 8MB qspi plus 16MB externally. Do you know which flash is used by micropython?

That's my understanding. It looks like MicroPython has implemented custom partition tables and it's making use of that. It almost looks like they have it set for 16mb internal flash. I attempted to set it to 16mb and compile, but the file system was still the same size.

**Comment by tannewt** [comment]
How do I flash the bin to the board? I'm not sure what bootloader it has.

**Comment by dhalbert** [comment]
> How do I flash the bin to the board? I'm not sure what bootloader it has.

It comes with some bootloader that pulses when reset is double-clicked. But I was not able to load anything with that bootloader
.
[MicroPython says](https://micropython.org/download/ARDUINO_NANO_ESP32/) to use `dfu-util`, but this did not work for me. It said there was no DFU device.

I also tried uploading Blink from the Arduino IDE (2.x.x) and that did not work either.

An Arduino Forum user [told](https://forum.arduino.cc/t/success-installed-micropython-in-new-nano-esp32/1152724) of a difficult experience: 

I finally found this:
https://support.arduino.cc/hc/en-us/articles/9625819325212-About-the-B1-and-B0-pins-on-the-Nano-ESP32

So short B1 to ground (the white-marked pin next to B1) and press reset. This puts it in regular Espressif low-level bootloader mode. Then I cleared flash (which presumably erased that pulsing bootloader):
```
esptool.py -p /dev/ttyACM0 erase_flash
```
pressed reset again, and loaded CircuitPython:
```
esptool.py -p /dev/ttyACM0 -b 921600 --before=default_reset --after=no_reset write_flash 0x0000 adafruit-circuitpython-arduino_nano_esp32s3-en_US-20230725-aaae963.bin
```
Presumably Web ESPTool would work too.

**Comment by makermelissa** [comment]
Hi sorry. yeah, as Dan said, you need to short the B1 pin to ground and flash via esptool or web serial esptool.

**Review by tannewt — CHANGES_REQUESTED** [review]
Ok, I've got it loaded. One thing I noticed is that the D13 LED is always on. I suspect this is due to the default pin state. You'll want to [override it in boards.c](https://github.com/adafruit/circuitpython/blob/305303996dd0a91ae87a82e76b4c0b8c9872859a/ports/espressif/boards/adafruit_feather_esp32_v2/board.c#L34)

The safe mode blink on mine was red instead of yellow and I'm not sure why. I can do yellow on the LED myself through code.py.

**Comment by makermelissa** [comment]
> The safe mode blink on mine was red instead of yellow and I'm not sure why. I can do yellow on the LED myself through code.py.

Updated. If it's not the right color and you're using the regular one, you may have to try the inverted one. Likewise if you were already using the inverted one.

File: ports/espressif/boards/arduino_nano_esp32s3/board.c [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    #ifdef DEBUG
+    common_hal_never_reset_pin(&pin_GPIO43);
+    common_hal_never_reset_pin(&pin_GPIO44);
+    #endif
+}
+
+bool espressif_board_reset_pin_number(gpio_num_t pin_number) {
+    if (pin_number == 13) {
+        // Set D13 LED to output by default.
+        gpio_set_direction(pin_number, GPIO_MODE_DEF_OUTPUT);
```
Comment (tannewt) [inline ports/espressif/boards/arduino_nano_esp32s3/board.c:43]: I think you'll want it as input with a pull down. (It is pull up by default.) As-is, you are switching to output without setting whether it is high or low. That'll be unpredictable.

**Review by tannewt — CHANGES_REQUESTED** [review]
One further suggestion.

File: ports/espressif/boards/arduino_nano_esp32s3/board.c (hunk 1 above)
Comment (makermelissa) [reply ports/espressif/boards/arduino_nano_esp32s3/board.c:43]: I'm confused. Shouldn't an LED be an output?

Comment (tannewt) [reply ports/espressif/boards/arduino_nano_esp32s3/board.c:43]: Only when in use. This function sets its state when not in use.

**Comment by makermelissa** [comment]
Sorry for the delay, but requested changes have been completed.

**Review by tannewt — APPROVED** [review]
Thank you!

**Pull request merged**